---
output: github_document
---


```{r, include=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  fig.bg = "transparent",
  out.width = "100%"
)
```


# **[SNPshiftStop](https://github.com/darkghastful/SNPshiftStop)**
<!-- Uses details for a single nucleotide polymorphisms (SNPs) and NCBI API queries to identify the premature stop within the encoded protein, sequences of the reference/altered DNA, and sequences of the reference/altered proteins. -->

Whole genome sequencing (WGS) is a process to identify single nucleotide polymorphisms (SNPs) or base pair differences in the DNA of an individual. Certain SNPs, single or double insertions/deletions within the exon of a protein coding sequence cause a frameshift. This is because three base pairs encode for a single amino acid, the insertion/deletion of base pairs will cause the three base pairs encoding for each amino acid to be altered downstream. The resulting protein is likely to be shortened and malformed which can be associated with disease or disease resistance. Notable examples of this are BRCA1/BRCA2 and CCR5. [Hereditary Breast and Ovarian Cancer](https://www.ncbi.nlm.nih.gov/books/NBK1247) is associated with mutations of the BRCA1 or BRCA2 gene resulting in increased risk of cancer. While a [32 base pair deletion](https://www.science.org/doi/abs/10.1126/science.273.5283.1856) in the [CCR5](https://www.nature.com/articles/381667a0) gene is associated with a resistance to HIV. Both of these are due to shortened disfunctional proteins. Knowledge of the exact location of the frameshift and the altered downstream sequence of a protein is crucial to understand how much functionality a protein will retain.

To accomplish this, users can call SNPshiftsearch with details for an SNP. The function queries NCBI for the protein coding sequence, introduces the mutation to the DNA, translates the DNA of the reference/alternate sequences to the protein sequences, identifies the location of the premature stop (if one was introduced by the mutation), and calculates the length of the shortened protein.

---

## **Installation**

<!-- From CRAN (when published) -->
<!-- install.packages("") -->

```{r, results = "hide", warning = FALSE, message = FALSE}
# install.packages("devtools")      
install.packages("darkghastful/SNPshiftStop")
```

---

## **Quick Start**

### **Load the package**

```{r, results = "hide", warning = FALSE, message = FALSE}
library(SNPshiftStop)
```

### **Usage example**

```{r, results = "hide", warning = FALSE, message = FALSE, eval=FALSE}
BRCA1.E75V <- SNPshiftStop("BRCA1", mutation.from.start=224:227, ref.bp="AAAG", ref.aa="E", alt.aa="V", aa.mut.pos=75)
```

Alternatively an rsid can be queried and the row containing the desired mutation can be passed to SNPshiftstop
```{r, results = "hide", warning = FALSE, message = FALSE, eval=FALSE}
CCR5.rs333 <- SNPsearch("rs333")
CCR5.S185I <- CCR5.rs333[4,]
CCR5.S185I <- SNPshiftStop(CCR5.S185I)
```


---

## **Functions**

### `SNPsearch(rsid)`

Uses an rsid to query NCBI and identifies all associated SNPs.

- **Arguments**  
  - `rsid` — Rsid associated with SNP.
- **Returns**  
  - A frame of SNPs.
  
### **Example**
#### `SNPsearch("rs333")`

```{r, fig.width=12, fig.asp=90/212, dpi=300, echo=FALSE}
SNPsearch("rs333")
```

### `SNPshiftStop(gene.symbol, taxon="human", mutation.from.start=NA, ref.bp="-", alt.bp="-", ref.aa=NA, alt.aa=NA, aa.mut.pos=NA, start.loss=FALSE, ncbi.api.key=NULL)`

Provides information on an SNP. 

- **Arguments**  
  - `gene.symbol` - Results of SNPsearch or gene symbol 
  - `taxon` - provide taxon character string or number; found using NCBI (default human (9606))
  - `mutation.from.start` - Location of mutation relative to start or
  - `ref.bp` - reference base pair (default is '"-"')
  - `alt.bp` - alternate base pair (default is '"-"')
  - `ref.aa` - reference amino acid (default is NA)
  - `alt.aa` - alternate amino acid (default is NA)
  - `aa.mut.pos` - amino acid mutation position (default is NA)
  - `start.loss` - does this mutation contain a start.loss (default is TRUE)
  
- **Returns**  
  A summary of the results will be printed including the reference amino acid sequence and alternate amino acid sequence.
  - `gene.symbol` - Gene symbol.
  - `aa.seq` - Amino acid sequence of the protein. 
  - `alternate.aa.seq` - Alternate amino acid sequence.
  - `shortened.of.total.aa` - Length of alternate amino acid sequence/length of original.
  - `dna.seq` - DNA sequence encoding the protein.
  - `alternate.dna.seq` - Alternate DNA sequence.
  
### **Example**
#### `SNPshiftStop("BRCA1", mutation.from.start=224:227, ref.bp="AAAG", ref.aa="E", alt.aa="V", aa.mut.pos=75)`

```{r scale, fig.width=12, fig.asp=90/212, dpi=300, echo=FALSE}
CFTR.F405L <- SNPshiftStop("BRCA1", mutation.from.start=224:227, ref.bp="AAAG", ref.aa="E", alt.aa="V", aa.mut.pos=75)
```





<!-- snp.query -->

<!-- API documentation[https://clinicaltables.nlm.nih.gov/apidoc/snps/v3/doc.html]: -->
<!-- Output for an API query is an array of the following elements: -->

<!-- - The total number of results on the server, which can be more than the number of results returned. This reported total number of results may also be significantly less than the actual number of results and is limited to 10,000, which may significantly improve the service response time. -->
<!-- - An array of codes for the returned items. (This is the field specified with the cf query parameter above.) -->
<!-- - A hash of the "extra" data requested via the "ef" query parameter above. The keys on the hash are the fields (or their requested aliases) named in the "ef" parameter, and the value for a field is an array of that field's values in the same order as the returned codes. -->
<!-- - An array, with one element for each returned code, where each element is an array of the display strings specified with the "df" query parameter. -->
<!-- - An array, with one element for each returned code, where each element is the "code system" for the returned code. Note that only code-system aware APIs will return this array. -->

---

## **Dependencies**

- **bqutils**  
- **Biostrings**
- **stringr**  
- **magrittr**  
- **methods**


---

## **License**

GPL-3.0 license © Bailey Quinn
